core/controller/src/main/scala/whisk/core/controller/actions/PrimitiveActions.scala

```diff
@@ -139,6 +199,356 @@ protected[actions] trait PrimitiveActions {
+      })
+
+    invokeConductor(user, payload, session)
+
+    // is caller waiting for the result of the activation?
+    waitForResponse
+      .map { timeout =>
+        // handle timeout
+        session.result.head.future
+          .withAlternativeAfterTimeout(timeout, Future.successful(Left(session.activationId)))
+      }
+      .getOrElse {
+        // no, return the session id
+        Future.successful(Left(session.activationId))
+      }
+  }
+
+  /**
+   * A method that knows how to handle a conductor invocation.
+   *
+   * This method prepares the payload and invokes the conductor action.
+   * It parses the result and extracts the name of the next component action if any.
+   * It either invokes the desired component action or completes the composition invocation.
+   * It also checks the invocation counts against the limits.
+   *
+   * @param user the identity invoking the action
+   * @param payload the dynamic arguments for the activation
+   * @param session the session object for this composition
+   * @param transid a transaction id for logging
+   */
+  private def invokeConductor(user: Identity, payload: Option[JsObject], session: Session)(
+    implicit transid: TransactionId) = {
+
+    if (session.accounting.conductors > 2 * actionSequenceLimit) {
+      // composition is too long
+      val response = ActivationResponse.applicationError(compositionIsTooLong)
+      completeAppActivation(user, session, response)
+    } else {
+      // inject state into payload if any
+      val params = session.state
+        .map { state =>
+          Some(JsObject(payload.getOrElse(JsObject()).fields ++ state.fields))
+        }
+        .getOrElse(payload)
+
+      // invoke conductor action
+      session.accounting.conductors += 1
+      val response =
+        invokeSimpleAction(
+          user,
+          session.action,
+          params,
+          Some(session.action.limits.timeout.duration + 1.minute), // wait for result
+          Some(session.activationId)) // cause is session id
+
+      response.onComplete {
+        case Failure(t) =>
+          // invocation failure
+          val response = ActivationResponse.whiskError(compositionActivationFailure)
+          completeAppActivation(user, session, response)
+        case Success(Left(activationId)) =>
+          // invocation timeout
+          session.logs += activationId
+          val response = ActivationResponse.whiskError(compositionActivationTimeout(activationId))
+          completeAppActivation(user, session, response)
+        case Success(Right(activation)) =>
+          // successful invocation
+          session.logs += activation.activationId
+          session.duration += activation.duration.getOrElse(activation.end.toEpochMilli - activation.start.toEpochMilli)
+
+          val result = activation.resultAsJson
+
+          // extract params from result
+          val params = result.getFields(WhiskActivation.paramsField).headOption.map { p =>
```

Comment:
should be same as 
` +          val params = result.fields.get(WhiskActivation.paramsField).map { p => `

here and elsewhere.

Reply:
OK will fix throughout.